lib.app.sciond unit tests

shitz opened the pull request:
Unit tests for `lib.app.sciond.py` (#986)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion/1020)
<!-- Reviewable:end -->

kormat commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[lib/app/sciond.py, line 208 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvuZxb6IGXzxY-SgnW:-KfvuZxb6IGXzxY-SgnX:b2qbd5n) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L208)):*
> ```Python
>         return None
> 
>     def _resolve_dst_addr(self, src, dst):  # pragma: no cover
> ```

This is complex enough that it should have unit tests.

---

*[lib/sciond_api/host_info.py, line 69 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvurHuTVs3uiub2Ang:-KfvurHuTVs3uiub2Anh:b-4bkkiw) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/sciond_api/host_info.py#L69)):*
> ```Python
>                 (self.ipv4() or "unset", self.ipv6() or "unset", self.p.port))
> 
>     def __eq__(self, other):
> ```

Where is this used?

---

*[test/lib/app/sciond_test.py, line 54 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvwOD81NBjNUGa-JtJ:-KfvwOD81NBjNUGa-JtK:bdtgu68) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L54)):*
> ```Python
>         connector = self._setup_connector(
>             self._create_response(["1", "2"], False))
>         # Call
> ```

It would be good to test that `SCIONDPathRequest.from_values` is being called correctly with the supplied parameters.

---

*[test/lib/app/sciond_test.py, line 65 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvwcqGiXs4DLHsAua6:-KfvwcqGiXs4DLHsAua7:bmqgd90) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L65)):*
> ```Python
>             self._create_response([], True))
>         # Call
>         with ntools.assert_raises(SCIONDResponseError):
> ```

`ntools.assert_raises(SCIONDResponseError, connector.get_paths, 1, 2, 5)`

---

*[test/lib/app/sciond_test.py, line 68 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-Kfvwj9sP2kVl9pi0BJC:-Kfvwj9sP2kVl9pi0BJD:b4aup15) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L68)):*
> ```Python
>             connector.get_paths(1, 2, 5)
>         connector._create_socket.assert_called_once_with()
>         connector._get_response.assert_called_once_with(ANY, 1, SMT.PATH_REPLY)
> ```

These two checks aren't needed, as they're covered by the first case.

---

*[test/lib/app/sciond_test.py, line 81 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvyNJLnHaP8_ufd1S8:-KfvyNJLnHaP8_ufd1S9:b-nttyqw) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L81)):*
> ```Python
>         ntools.eq_(connector.get_as_info(), ["as_info"])
>         # Tests
>         ntools.eq_(connector._as_infos["local"], ["as_info"])
> ```

`_try_cache` and `SCIONASInfoRequest.from_values` calls aren't checked.

---

*[test/lib/app/sciond_test.py, line 108 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-Kfw2A4CvJw9Ab_I8NzZ:-Kfw2A4CvJw9Ab_I8Nz_:b33sjrw) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L108)):*
> ```Python
> 
> 
> class TestSCIONDConnectorGetIFInfo(SCIONDConnectorTestBase):
> ```

Looking at the complexity and number of these tests, it seems like an indicator that `get_if_info` should be refactored. The first half of it deals with cached entries. That could be split off into a smaller method. Something like:
```
def _get_cached_if_info(self, if_list):
  if_set = set(if_list)
  if not if_set:
    return set(), {}
  infos = self._try_cache(self._if_infos, if_set)
  if_set -= set(infos)
  return if_set, infos
```

`get_if_info` would then look like:
```
with ...lock:
  if_set, if_infos = self._get_cached_if_info(if_list)
  if not if_set:
    return if_infos
  .... # proceed with sending queries
```
  
  

---

*[test/lib/app/sciond_test.py, line 254 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-Kfw3xEvGbxqJtCN6bU8:-Kfw3xEvGbxqJtCN6bU9:b-z5s0il) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L254)):*
> ```Python
> 
> class TestSCIONDConnectorGetOverlayDest:
>     """Unit tests for lib.app.sciond.SCIONDConnector.get_overlay_dest"""
> ```

I would sooner mark `get_overlay_dest` as `no cover`, and have tests just for `_resolve_dst_addr`. Ensuring that `_resolve_dst_addr` is adequately tested via `get_overlay_dest` makes things complicated, and harder to check correctness. `get_overlay_dest` and `_resolve_ifid` are simple enough to be easily checked visually.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[lib/app/sciond.py, line 208 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvuZxb6IGXzxY-SgnW:-KfzpHpLLPITag-fMLpR:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L208)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is complex enough that it should have unit tests.
</blockquote></details>

Done.

---

*[lib/sciond_api/host_info.py, line 69 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvurHuTVs3uiub2Ang:-KfzFZOa0aFEOizOOHUo:bidw0oo) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/sciond_api/host_info.py#L69)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Where is this used?
</blockquote></details>

Was used in `test_with_host`. Mocked out now and removed.

---

*[test/lib/app/sciond_test.py, line 54 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvwOD81NBjNUGa-JtJ:-KfzJKUIQF2KSFzZuMIj:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L54)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

It would be good to test that `SCIONDPathRequest.from_values` is being called correctly with the supplied parameters.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 65 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvwcqGiXs4DLHsAua6:-KfzK6Co8985D_AihmFr:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L65)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`ntools.assert_raises(SCIONDResponseError, connector.get_paths, 1, 2, 5)`
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 68 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-Kfvwj9sP2kVl9pi0BJC:-KfzKGyCrpLandZTOBZl:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L68)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

These two checks aren't needed, as they're covered by the first case.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 81 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KfvyNJLnHaP8_ufd1S8:-KfzPQU_J1btRVugd3-6:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L81)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`_try_cache` and `SCIONASInfoRequest.from_values` calls aren't checked.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 108 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-Kfw2A4CvJw9Ab_I8NzZ:-KfzpNxED5uG6iq48MWR:bwva7y1) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L108)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Looking at the complexity and number of these tests, it seems like an indicator that `get_if_info` should be refactored. The first half of it deals with cached entries. That could be split off into a smaller method. Something like:
```
def _get_cached_if_info(self, if_list):
  if_set = set(if_list)
  if not if_set:
    return set(), {}
  infos = self._try_cache(self._if_infos, if_set)
  if_set -= set(infos)
  return if_set, infos
```

`get_if_info` would then look like:
```
with ...lock:
  if_set, if_infos = self._get_cached_if_info(if_list)
  if not if_set:
    return if_infos
  .... # proceed with sending queries
```
  
  
</blockquote></details>

Moved this functionality to `_try_cache` and added unit tests for it.

---

*[test/lib/app/sciond_test.py, line 254 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-Kfw3xEvGbxqJtCN6bU8:-KfzpU_0HBLwSYNXkzio:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/test/lib/app/sciond_test.py#L254)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I would sooner mark `get_overlay_dest` as `no cover`, and have tests just for `_resolve_dst_addr`. Ensuring that `_resolve_dst_addr` is adequately tested via `get_overlay_dest` makes things complicated, and harder to check correctness. `get_overlay_dest` and `_resolve_ifid` are simple enough to be easily checked visually.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[lib/app/sciond.py, line 217 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUaidjfOn0WcBUwezm:-KgUaidjfOn0WcBUwezn:bss08uq) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L217)):*
> ```Python
>             svc_infos = self.get_service_info([svc_type])
>             if svc_type in svc_infos:
>                 return svc_infos[svc_type].host_info()
> ```

The dict that `get_service_info` returns only has a single entry for each service? What happens if it was a multicast SVC type requested?

---

*[lib/app/sciond.py, line 137 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgURmntZxP9YzhlOLFh:-KgURmntZxP9YzhlOLFi:b-ek77vb) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L137)):*
> ```Python
>             q_ia = "local"
>         with self._as_infos_lock:
>             as_info = self._try_cache(self._as_infos, [q_ia])[1].get(q_ia)
> ```

This is a bit complex for a single expression, i suggest splitting into two expressions.

---

*[lib/app/sciond.py, line 246 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQSPGui1_cgrIUHak:-KgUQSPGui1_cgrIUHal:bvellq1) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L246)):*
> ```Python
> 
>     @staticmethod
>     def _try_cache(cache, key_list):
> ```

This needs a docstring saying what it does (and what it returns).

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-KgUQuRzMIlIrHVDJV7I:b-mgb5mn) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
> ```Python
>         key_set = set(key_list) if key_list else set()
>         if not key_set:
>             return set(), {}
> ```

I don't follow this logic. E.g. `get_if_info` calls `_try_cache` with either a list of IF IDs, or `None`. If it's `None`, (or the list is empty) the cache isn't consulted at all? Looking back at the previous version, that does seem to be the existing logic, but it doesn't look right.

---

*[test/lib/app/sciond_test.py, line 56 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUThE8ZNS1HRX-UB_d:-KgUThE8ZNS1HRX-UB_e:bgq1osl) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L56)):*
> ```Python
> 
>     @patch("lib.app.sciond.SCIONDPathRequest.from_values", new_callable=create_mock)
>     def test(self, from_values):
> ```

`path_req` would be a better name.

---

*[test/lib/app/sciond_test.py, line 67 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUTnRao0-2WTWh7r0D:-KgUTnRao0-2WTWh7r0E:bbeo9ho) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L67)):*
> ```Python
>         connector._create_socket.assert_called_once_with()
>         connector._get_response.assert_called_once_with(ANY, 1, SMT.PATH_REPLY)
>         from_values.assert_called_once_with(
> ```

This should move before the `_create_socket` check, as that's the execution order.

---

*[test/lib/app/sciond_test.py, line 84 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUUL2MmNNS-U8UpKQV:-KgUUL2NA9rx0WQJuxvm:b-f28x5r) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L84)):*
> ```Python
> 
>     @patch("lib.app.sciond.SCIONDASInfoRequest.from_values", new_callable=create_mock)
>     def test_local(self, from_values):
> ```

`as_info_req`

---

*[test/lib/app/sciond_test.py, line 93 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUVYM3zcweVXf36jNP:-KgUVYM3zcweVXf36jNQ:blhvb0q) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L93)):*
> ```Python
>         connector._get_response.assert_called_once_with(ANY, 1, SMT.AS_REPLY)
>         connector._try_cache.assert_called_once_with(connector._as_infos, ["local"])
>         from_values.assert_called_once_with(self.REQ_ID, None)
> ```

Execution order vs test order.

---

*[test/lib/app/sciond_test.py, line 143 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUZnDdkcfHSZpKo5uI:-KgUZnDdkcfHSZpKo5uJ:b-eq2o94) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L143)):*
> ```Python
> 
>     @patch("lib.app.sciond.SCIONDIFInfoRequest.from_values", new_callable=create_mock)
>     def test_get(self, from_values):
> ```

`test_get_without_cache` is a better name.

---

*[test/lib/app/sciond_test.py, line 281 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUbHpYH7BWKqCCYrmT:-KgUbHpYH7BWKqCCYrmU:bymjctx) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L281)):*
> ```Python
> 
>     def test_with_svc(self):
>         dst_addr = SCIONAddr.from_values(ISD_AS("1-1"), HostAddrSVC(0, raw=False))
> ```

`lib.packet.svc.SVCType.BS_A` and `lib.packet.svc.SVCType.BS_M` exist.

---

*[test/lib/app/sciond_test.py, line 294 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUd7XhnMlrZppwNLUY:-KgUd7XhnMlrZppwNLUZ:b-aek6ok) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L294)):*
> ```Python
>         src_addr = SCIONAddr.from_values(ISD_AS("1-1"), HostAddrIPv4("127.0.0.1"))
>         connector = self._setup_connector()
>         from_values.return_value = (dst_addr, SCION_UDP_EH_DATA_PORT)
> ```

You don't need to set an explicit return value. You can simply use `from_values.return_value` in the Call check below, without setting it to anything. The actual value doesn't matter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[lib/app/sciond.py, line 217 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUaidjfOn0WcBUwezm:-Kg_3MBEo9Pz9NIQX6PD:bw9za3v) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L217)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

The dict that `get_service_info` returns only has a single entry for each service? What happens if it was a multicast SVC type requested?
</blockquote></details>

That is a good catch. It's not a problem of the dictionary. A `SCIONDServiceInfoReplyEntry` has a list of `HostInfo` objects. Fixed to return the first one of those.

---

*[lib/app/sciond.py, line 137 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgURmntZxP9YzhlOLFh:-Kg_2nrvh9LwpbPI8GKE:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L137)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is a bit complex for a single expression, i suggest splitting into two expressions.
</blockquote></details>

Done.

---

*[lib/app/sciond.py, line 246 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQSPGui1_cgrIUHak:-Kg_4_Zyoy5zuNPaOkSm:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L246)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This needs a docstring saying what it does (and what it returns).
</blockquote></details>

Done.

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-Kg_4fHg3JqWaXWJ5pOc:b-vfq3md) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I don't follow this logic. E.g. `get_if_info` calls `_try_cache` with either a list of IF IDs, or `None`. If it's `None`, (or the list is empty) the cache isn't consulted at all? Looking back at the previous version, that does seem to be the existing logic, but it doesn't look right.
</blockquote></details>

How can you know that you have ALL IF infos in the cache when the list of IF IDs is None? In that case you have to ask SCIOND about all IF infos, thus you might as well not consultate the cache.

---

*[test/lib/app/sciond_test.py, line 56 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUThE8ZNS1HRX-UB_d:-Kg_5sidbguf72HSQDdq:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L56)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`path_req` would be a better name.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 67 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUTnRao0-2WTWh7r0D:-Kg_6naW-MotLON-0B3l:b-jpryir) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L67)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This should move before the `_create_socket` check, as that's the execution order.
</blockquote></details>

Damn you and your pickiness :P

---

*[test/lib/app/sciond_test.py, line 84 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUUL2MmNNS-U8UpKQV:-Kg_708XU9fJZG1ShdgJ:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L84)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`as_info_req`
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 93 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUVYM3zcweVXf36jNP:-Kg_6tv8Ptva8vYnAhpK:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L93)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Execution order vs test order.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 143 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUZnDdkcfHSZpKo5uI:-Kg_7F3frpiIY8ZWBtV2:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L143)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`test_get_without_cache` is a better name.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 281 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUbHpYH7BWKqCCYrmT:-Kg_7Zd1AP2t0giBBbHd:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L281)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`lib.packet.svc.SVCType.BS_A` and `lib.packet.svc.SVCType.BS_M` exist.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 294 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUd7XhnMlrZppwNLUY:-Kg_7lpS0maaSzLz-Sjt:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L294)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

You don't need to set an explicit return value. You can simply use `from_values.return_value` in the Call check below, without setting it to anything. The actual value doesn't matter.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[lib/app/sciond.py, line 217 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUaidjfOn0WcBUwezm:-KgnXucJQQmfAjbvWNEa:b-jbyts) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L217)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

That is a good catch. It's not a problem of the dictionary. A `SCIONDServiceInfoReplyEntry` has a list of `HostInfo` objects. Fixed to return the first one of those.
</blockquote></details>

I'm not sure if it makes sense to return a single value for an MSVC address, but i'm also not sure if it's valid for a client of sciond to use MSVC either.

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-KgnAy7U9lnuI_IuC0td:bc6lgp5) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

How can you know that you have ALL IF infos in the cache when the list of IF IDs is None? In that case you have to ask SCIOND about all IF infos, thus you might as well not consultate the cache.
</blockquote></details>

In that case it would be much clearer to check if that's the case, and not call `_try_cache` at all. Also the docstring says it returns a list of keys not found, but that's not true at all for the `None` case.

---

*[test/lib/app/sciond_test.py, line 67 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUTnRao0-2WTWh7r0D:-KgnYNxK5W5RYicobGdn:b-j2upi3) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L67)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Damn you and your pickiness :P
</blockquote></details>

You try reviewing unit test coverage sometime, you'll quickly understand. I have scars.
(Also you didn't fix this one).

---

*[test/lib/app/sciond_test.py, line 93 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUVYM3zcweVXf36jNP:-KgnYvX8N3uup7Rz1-7C:b6ld5te) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L93)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`_try_cache` is the first thing called, then `from_values`, then `_create_socket`, then `_get_response`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[lib/app/sciond.py, line 217 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUaidjfOn0WcBUwezm:-Kgnca0pm58ykRLTUVVx:bgqvsms) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L217)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I'm not sure if it makes sense to return a single value for an MSVC address, but i'm also not sure if it's valid for a client of sciond to use MSVC either.
</blockquote></details>

So what do you suggest?

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-Kgne63N8LLYPSlaYPDR:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

In that case it would be much clearer to check if that's the case, and not call `_try_cache` at all. Also the docstring says it returns a list of keys not found, but that's not true at all for the `None` case.
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 67 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUTnRao0-2WTWh7r0D:-KgneQmjCYWl6UZKn2D3:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L67)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

You try reviewing unit test coverage sometime, you'll quickly understand. I have scars.
(Also you didn't fix this one).
</blockquote></details>

Done.

---

*[test/lib/app/sciond_test.py, line 93 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUVYM3zcweVXf36jNP:-KgnepUcR2e4-HP2Yq_q:b9fd2ge) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/test/lib/app/sciond_test.py#L93)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`_try_cache` is the first thing called, then `from_values`, then `_create_socket`, then `_get_response`.
</blockquote></details>

Done and fixed some other occurrences of this as well. Hope I got all of them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/app/sciond.py, line 217 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUaidjfOn0WcBUwezm:-KgoMSG-ZvDd0GouXALa:brg8bmw) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L217)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

So what do you suggest?
</blockquote></details>

Probably the best thing for now is to return an error if an MSVC address is requested. That way at least we don't miss unexpected usage of a case we don't support.

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-KgoMLsTYs45KfXbD8Z6:bjgrour) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The code is still here in `_try_cache`.

---

*[lib/app/sciond.py, line 158 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgoMekde1bEyG9iwj1r:-KgoMekde1bEyG9iwj1s:bbfgjbd) ([raw file](https://github.com/netsec-ethz/scion/blob/508474c6d1f1d3eb4972ca25f7274f4b27b31fc7/lib/app/sciond.py#L158)):*
> ```Python
>                 if_infos = {}
>                 if_list = set()
>             if not if_list and if_infos:
> ```

If this is moved to directly after the `_try_cache` call, then you don't need to test `if_infos`.
```
if_list, if_infos = self._try_cache(...)
if not if_list:
  return if_infos
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

shitz commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[lib/app/sciond.py, line 217 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUaidjfOn0WcBUwezm:-Kgt7iMPwKJ2HpneMRuT:btd8fu2) ([raw file](https://github.com/netsec-ethz/scion/blob/eab488167192cc07c2bea933969668e9f46770e9/lib/app/sciond.py#L217)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Probably the best thing for now is to return an error if an MSVC address is requested. That way at least we don't miss unexpected usage of a case we don't support.
</blockquote></details>

After offline discussion we decided to leave it as it is. This API call will handle MSVC addresses just like ASVC addresses. Should we ever see the need to support MSVC addresses for client, we'd implement it in a different library call.

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-KgtBUJtIqyavXGQzER8:b16ptdc) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

The code is still here in `_try_cache`.
</blockquote></details>

Removed, but now the app will crash should `key_list` ever be `None` (which should never happen since we control all the code that calls it, but still)

---

*[lib/app/sciond.py, line 158 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgoMekde1bEyG9iwj1r:-Kgt6KJf4R7gW7Fj5U8v:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/508474c6d1f1d3eb4972ca25f7274f4b27b31fc7/lib/app/sciond.py#L158)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

If this is moved to directly after the `_try_cache` call, then you don't need to test `if_infos`.
```
if_list, if_infos = self._try_cache(...)
if not if_list:
  return if_infos
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020)*
<!-- Sent from Reviewable.io -->

kormat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/app/sciond.py, line 249 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-KgUQuRzMIlIrHVDJV7H:-Kgya-bc3YgxFnEZj8aP:bk4mix) ([raw file](https://github.com/netsec-ethz/scion/blob/5810f13c07baecb0783a41a5b235bac55f67d7a6/lib/app/sciond.py#L249)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

Removed, but now the app will crash should `key_list` ever be `None` (which should never happen since we control all the code that calls it, but still)
</blockquote></details>

This is desirable (IMO), as it means we'll immediately know about a bug.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1020#-:-Kgya6IgYseRrMRaFXOA:bnfp4nl)*
<!-- Sent from Reviewable.io -->
